File: Chapter 08/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter08/section_8_1_4_database_comparison/infrastructure/NoSqlSimulator.java
// infrastructure/NoSqlSimulator.java
package com.grokkingsoftwarearchitecture.chapter08.section_8_1_4_database_comparison.infrastructure;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * THE DOCUMENT WAY (INFRASTRUCTURE LAYER): Fast, loose. Like a messy desk.
 * Simulates MongoDB's document storage using standard Java lists and objects.
 */
public class NoSqlSimulator {
    public final List<NoSqlDocument> collection = new ArrayList<>();

    public void insertOne(NoSqlDocument document) {
        collection.add(document);
    }

    // The naive literal search
    public List<String> findByName(String name) {
        return collection.stream()
                .filter(doc -> doc.name.equals(name))
                .map(doc -> doc.name)
                .collect(Collectors.toList());
    }

    public List<String> findByTag(String tag) {
        // Contains Match: Better, but still relies on exact spelling of the tag.
        return collection.stream()
                .filter(doc -> doc.tags.contains(tag))
                .map(doc -> doc.name)
                .collect(Collectors.toList());
    }
}

File: Chapter 08/C#/Section_8_1_4_DB_Comparison/Infrastructure/NoSqlSimulator.cs
// Infrastructure/NoSqlSimulator.cs
using System.Collections.Generic;
using System.Linq;

namespace Chapter08.DatabaseCode.Infrastructure
{
    /// <summary>
    /// THE DOCUMENT WAY (INFRASTRUCTURE LAYER): Fast, loose. Like a messy desk.
    /// Simulates MongoDB's document storage using standard C# lists and objects.
    /// </summary>
    public class NoSqlSimulator
    {
        public readonly List<NoSqlDocument> Collection = new List<NoSqlDocument>();

        public void InsertOne(NoSqlDocument document)
        {
            Collection.Add(document);
        }

        // The naive literal search
        public List<string> FindByName(string name)
        {
            return Collection
                .Where(doc => doc.Name == name)
                .Select(doc => doc.Name)
                .ToList();
        }

        public List<string> FindByTag(string tag)
        {
            // Contains Match: Better, but still relies on exact spelling of the tag.
            return Collection
                .Where(doc => doc.Tags.Contains(tag))
                .Select(doc => doc.Name)
                .ToList();
        }
    }
}